### PR TITLE
Fix ITableDataSource support for multiple sinks

### PR DESCRIPTION
**Pull request by AArnott**
While the VS TDS only adds one sink of its own, this is something of an implementation detail, and doesn't support anyone else adding their own sink, which would lead to this code malfunctioning.
With this change, the FindUsages code tracks and updates all sinks, and only terminates an async search when *all* sinks have been removed.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 1]
```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
...
+            {
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Remove(arg), sink);
 
-                CancelSearch();
+                if (_tableDataSinks.IsEmpty)
+                {
+                    // TODO: Why is this a safe assumption? Where is the contract that these subscriptions are only disposed on the main thread?
+                    //       Even if VS always does it, other folks can add their own sinks, and if the requirement isn't documented,
+                    //       others may dispose on another thread.
+                    Presenter.AssertIsForeground();
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: This method is the one I'm least confident in. Which of these operations should happen when *all* sinks are removed vs. perhaps the removal of just one (specially recognized?) sink?

Also of course is my TODO comment about why this is assumed to be on the UI thread, and whether we can remove that requirement since I don't think that's a guarantee or requirement on the interface.

**Comment by CyrusNajmabadi** [comment]
I'm not really sure why we need to do this.  When we originally wrote this, we talked to editor and got the guarantee from them. 

We also control everything here.  We control the code that adds the source to the manager and then gets informed by the manager about the sink.  So this all seems internal and doesn't need any logic to support things that should never happen.

--

Note: if we really are concerned about someone doing the wrong thing, we could just make the assert in 'Subscribe' into a throw.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 2]
```diff
@@ -27,12 +27,17 @@ private abstract class AbstractTableDataSourceFindUsagesContext :
...
+            private ImmutableList<ITableDataSink> _tableDataSinks = ImmutableList<ITableDataSink>.Empty;
 
             public readonly StreamingFindUsagesPresenter Presenter;
             private readonly IFindAllReferencesWindow _findReferencesWindow;
             protected readonly IWpfTableControl2 TableControl;
 
+            /// <summary>
+            /// Indicates the last stable state communicated with sinks.
+            /// </summary>
+            private bool isStable;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: private fields should be prefaced with `_`.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 3]
```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
...
+                private readonly AbstractTableDataSourceFindUsagesContext _owner;
+                private readonly ITableDataSink _sink;
+
+                internal DisposableSubscription(AbstractTableDataSourceFindUsagesContext owner, ITableDataSink sink)
+                {
+                    _owner = owner ?? throw new ArgumentNullException(nameof(owner));
+                    _sink = sink ?? throw new ArgumentNullException(nameof(sink));
+                }
+
+                public void Dispose() => _owner.RemoveSink(_sink);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: please place `=>` on the next line, indented.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: this was communicated by the editor team as to how this will behave.  ANd our assert checks this (inlcuding in Release builds iirc) to verify this so that if that contract is violated we can inform them immediately.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 4]
```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
...
-                // of some other event (like the solution being closed).  Remove us from the set
-                // of sources for the window so that the existing data is cleared out.
-                Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 1);
-                Debug.Assert(_findReferencesWindow.Manager.Sources[0] == this);
+            #endregion
 
-                _findReferencesWindow.Manager.RemoveSource(this);
+            private void RemoveSink(ITableDataSink sink)
+            {
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Remove(arg), sink);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: please don't use Interlocked here. We have UI thread affinitized data that should just use normal constructs, and we have free-threaded data that should be protected by _gate.  This mixes concepts in terms of how data should be operated upon.

**Comment by CyrusNajmabadi** [comment]
> and doesn't support anyone else adding their own sink, which would lead to this code malfunctioning.

A better way to avoid that would be to make our 'Context' objects not be subclasses of FindRefsContxt, but instead be nested-private classes. Then, no one else could even get access to the instance to call .Subscribe on.  

I'm strongly against thsi change as we should *never* have multiple sinks.  And, if we're concerned about the off chance someone might do that, we should just eliminate that from being possible in the first place, thus allowing us to know we can only have 1 sink, instead of trying to make the code support something that should never happen.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 1 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: I am happy to accept that, but the docs on the interface should be updated to indicate that the interface must only be called from the UI thread.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 2 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Oh right. done.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 3 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Yup. Another style I missed. Thanks.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: it's also not clear how this is ever read... did i miss that somewhere?

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 5]
```diff
@@ -262,13 +272,12 @@ public IDisposable Subscribe(ITableDataSink sink)
...
                 Presenter.AssertIsForeground();
 
-                Debug.Assert(_tableDataSink == null);
-                _tableDataSink = sink;
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Add(arg), sink);
 
-                _tableDataSink.AddFactory(this, removeAllFactories: true);
-                _tableDataSink.IsStable = false;
+                sink.AddFactory(this, removeAllFactories: true);
+                sink.IsStable = this.isStable;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Gotcha.  That makes a lot of sense.  This needs to be documented in the code.  Either on the field, or at the read/write points (or both places).  It's def something that needs to made clear.  Thanks! :)

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs [hunk 6]
```diff
@@ -286,7 +295,11 @@ public sealed override async Task OnCompletedAsync()
             {
                 await OnCompletedAsyncWorkerAsync().ConfigureAwait(false);
 
-                _tableDataSink.IsStable = true;
+                this.isStable = true;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: ignore this thread.  we can just dicuss things in https://github.com/dotnet/roslyn/pull/32136#pullrequestreview-189233781

**Comment by CyrusNajmabadi** [comment]
> In fact we are in communication with the Roslyn team about the scenarios driving this change. :) But I'm not at liberty to discuss it openly. So while I appreciate your code review for correctness, I think we need to reserve the why and direction questions to current FTEs. sorry.

This seems highly against hte spirit and goals of Roslyn.  Being able to effectively decide the right ways to handle such changes without even having an clue about the purpose of the change is pretty difficult.  

Regardless, i would push back against this using the current interfaces.  You will break 3rd party plugins.  If you want to support a model where the end plugin isn't in control of the source/sink hookup, then we should add a new interface for that that the plugin can opt-into.  We can also consider if that is thread-affinitized or free-threaded.  If the latter, we can make the necessary changes in the impl to avoid race conditions that would now exist.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 4 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Yes, I agree. In fact as I was writing my own source table code, I found that forcing sequential handling of events was necessary to make sure my data integrity would be resilient to concurrent execution, etc.
That said, I'm not proposing any breaking changes here. AFAIK the table manager itself isn't changing so no old code should be broken. 
Assuming the scenario owner (Roslyn's feature at this point) keeps the table control / manager instance concealed from the outside world, they continue to be fine. But to support multiple sinks, it is likely that the table manager instance will have to be shared, which allows folks to enumerate the various sources and add their own sinks to them. So in preparation for Roslyn exposing the table manager (which would be a subsequent PR), we first need to prepare Roslyn for dealing with multiple sinks to avoid bugs. That's what this PR is about.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: code comment added.

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Sure.  But in this case, i would just keep the threading assert (first in the method), and just do a normal .Add or .Remove on the immutable-array/list.  The Interlocked serves no purpose.  Either we're UI thread only, in which case it's unnecessary.  Or we're free-threaded, in which case none of the code is corrrect, and being Interlocked doesn't change that.  Thanks! :)

**Comment by AArnott** [comment]
> Regardless, i would push back against this using the current interfaces. You will break 3rd party plugins.

Can you tell me more about that? What are 3rd party plugins, and how will these changes break them?

**Comment by CyrusNajmabadi** [comment]
> Can you tell me more about that? What are 3rd party plugins, and how will these changes break them?

All of these are public APIs.  So i mean: any third party that is using these public APIs to work with the ITableXXX system.  If you change how these work (i.e. adding more sinks, or changing threading) you may break them.  Imagine, for example, that Roslyn wasn't something you could just change when you needed (i.e. because Roslyn was a 3rd party), then this would be a problem.  Now consider all the other languages/provders/plugins that aren't 1st party.  How will they be affected?

File: src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs (hunk 4 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Fair enough. I'm removing the interlocks and asserting the ui thread everywhere.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs]: Thanks! :)

**Comment by CyrusNajmabadi** [comment]
FWIW, can you not force-push?  It just makes it hard to review.   Once we're done with teh code changes, you can either squash, rebase, force-push, whatever at that point :)

**Comment by AArnott** [comment]
> All of these are public APIs.

Which APIs are you referring to? I'm not changing any public APIs here. It's just the one file, which contains a private nested class. 
The relevant public APIs here are the interfaces that this class implements. The interface was clearly designed to *support* multiple sinks. Roslyn's implementation chose to be based on the assumption that there never would be multiple sinks. I'm adding support for it. What exactly would break?

> So i mean: any third party that is using these public APIs to work with the ITableXXX system. If you change how these work (i.e. adding more sinks, or changing threading) you may break them. 

So it doesn't sound like you're talking about Roslyn plugins then. Is that right? I can't see how anyone using the table interfaces would be broken from *this* change. We're expanding the set of use cases that Roslyn's implementation of an interface can support to be closer to what the interface was originally intended to support. And I daresay, I'm rather surprised it didn't implement support for it. If Roslyn *was* a 3rd party, closed source component, and I had a reference to a Roslyn object that implemented this interface, I would assume that they implemented the interface as it was designed and documented, which would allow multiple sinks.
Now granted, Roslyn hasn't exposed this object so there aren't any 3rd parties to Roslyn that can call into this code (yet). So again, I don't see how this can break anyone (barring bugs I may have introduced).

> Imagine, for example, that Roslyn wasn't something you could just change when you needed (i.e. because Roslyn was a 3rd party), then this would be a problem. Now consider all the other languages/provders/plugins that aren't 1st party. How will they be affected?

They won't be affected -- because I'm not making a breaking change. If Roslyn weren't OSS where I could send a PR, I'd simply file a bug against Roslyn asking that they expose their table manager and support multiple sinks so that we can add value to their scenario. And that's what I expect we'll eventually do to all the other language services out there, if they care to participate in what we're building.

**Comment by AArnott** [comment]
> FWIW, can you not force-push? It just makes it hard to review. Once we're done with teh code changes, you can either squash, rebase, force-push, whatever at that point :)

Ah, it's refreshing to hear someone else besides me advocate for this position. I rarely force push (and I *never* squash in a PR). But I've heard from so many repo owners that they want clean PRs with (nearly) one commit, so I've started force pushing to adopt their PR feedback rather than adding commits to get ahead of the criticism. But I have a healthy respect for just adding commits to refine code. I'm happy to oblige.

FWIW though, github *has* made it easier. With each force push, github adds an entry to the PR log that lets you see the before and after force-push commits. I think that would resolve your concern. Have you tried that?

**Comment by CyrusNajmabadi** [comment]
> The relevant public APIs here are the interfaces that this class implements. The interface was clearly designed to support multiple sinks. Roslyn's implementation chose to be based on the assumption that there never would be multiple sinks. 

It's not an assumption.  We own the creation of this type and we're the ones that call .AddSource on the table manager.  No one else should ever do that or be able to get access to the instance to do that.  

How could you ever get multiple sinks?

> I'm adding support for it. What exactly would break?

The breaking is that if there can now be multiple sinks (including other sinks coming in from some random location) this can break assumptions.  i.e. imagine a *3rd party* had written this:

```c#
            public IDisposable Subscribe(ITableDataSink sink)
            {
                if (_tableDataSink != null)
                     throw new InvalidOperationException("Can't happen!  We're the only one registering ourselves as the source!  BAD BAD BAD!");
```

This is now broken.

**Comment by CyrusNajmabadi** [comment]
>  But I've heard from so many repo owners that they want clean PRs with (nearly) one commit, 

Do they want the PRs to be one commit?  Or do they want the merge to just be squashed into a single commit at the end? :)

> FWIW though, github has made it easier. With each force push, github adds an entry to the PR log that lets you see the before and after force-push commits. I think that would resolve your concern. Have you tried that?

I've never found this easy or clear to use.  It always seems like i'm seeing changes that i've already seen.

**Comment by AArnott** [comment]
> imagine a 3rd party had written this:

This PR wouldn't break that 3rd party. I'm not adding a second sink in this PR to Roslyn's table source or anyone else's. So this PR can't break even those folks who, like Roslyn, assumed there would ever only be one sink.

Perhaps you're concerned not about *this* PR breaking others, but the change that you imagine will follow. If you're thinking that I'll be introducing code that will add sinks to arbitrary table sources in VS, then yes, I agree that could cause them to throw or otherwise malfunction. I'm not proposing to do that. My code will only add sinks to table sources that opt in. And *this* PR prepares Roslyn to be ready to opt in.

>> FWIW though, github has made it easier. With each force push, github adds an entry to the PR log that lets you see the before and after force-push commits. I think that would resolve your concern. Have you tried that?

> I've never found this easy or clear to use. It always seems like i'm seeing changes that i've already seen.

Are you actually clicking the "force-pushed" link? That only shows me the diff from one push to another.

**Comment by CyrusNajmabadi** [comment]
>  And I daresay, I'm rather surprised it didn't implement support for it.

It was a 100% explicit decision based on communication directly with Editor on this.  There simply was no way for this to happen.  I asked *directly* if this were possible/expected.  I also got crisp guidance on when these events would happen.  For example, i wanted to know if adding the source was guaranteed to callback immediately to subscribe the sink.  

Multiple sinks would represent a bug (because we own this entire end-to-end scenario and do not expose anything to anyone else to interact with), and i'm a strong proponent of not writing code that tries to be resilient to things that should not happen.  Instead, i want those cases to throw and for very explicit communication and discussion go into what the right fix was.

--

For example, say Roslyn had decided to support multiple sinks (even though it should not happen), and hten the editor introduced a *bug* by which they multiply subscribed (potentially only diposing one, but not all of the registrations).  I don't want this code to just blindly be ok with that, causing a leak.  I want this code to die immediately, saying: something very unexpected is happening, and it should nto be happening since we own the end-to-end here and have a strong understanding of how things should be used.

There are many scenarios where we're involved with other subsystems, but the interaction and lifetimes are very well understood and defined (though could probably be better in docs).  In these cases, i don't want our code to be written to work in case the other subsystems deviate from that understanding.  I want to code as tightly as possible, and then be very explicit about if/when we need to change things.

**Comment by CyrusNajmabadi** [comment]
>Are you actually clicking the "force-pushed" link? That only shows me the diff from one push to another.

Yes.  Though this is not pleasant in github.  Because instead of using the 'commit' tab, i now need to both use the commit-tab, and the conversation-tab.  And hte conversation tab is chock full of lots of back-and-forth between us, and trying to find these force-pushes is super annoying.  There are three of them in random locations on the page.  Instead of having a super clean 'commit tab' that i can walk through, i now need to jump back/forth annoyingly.

**Comment by CyrusNajmabadi** [comment]
> This PR wouldn't break that 3rd party. I'm not adding a second sink in this PR to Roslyn's table source or anyone else's

I'm not speaking about this PR in particular.  But just the general concept of another sink being added, implying someone else is intracting with the lifetimes of these subsystems out of the control of the plugin itself.  In Roslyn's case, we're the ones adding ourselves as the source to the table manager.  That should only cause a single Subscribe to happen.  If that now causes multiple changes in teh future that would break us (or anyone like us) that did not expect that to ever happen, and which we got communication would not happen.  

Basically, the lifetime and interaction of these calls was explained and at least verbally documented (i'm too lazy to go read the docs to see if there's anything there as well).  So changing that breaks expectations.  This is a comment about the *general overall change happening*, not necessarily this PR in particular.  It's just that this PR helps show a general problem that others will face.

> My code will only add sinks to table sources that opt in. And this PR prepares Roslyn to be ready to opt in.

I see.  Can we wait until that point?  I'd like to see how a plugin would opt-into this new behavior.  Thanks! :)

**Comment by AArnott** [comment]
> I see. Can we wait until that point? I'd like to see how a plugin would opt-into this new behavior. Thanks! :)

Ah good, I'm glad it's more clear now. :)

And as for waiting till then, sure. Your arguments for not supporting scenarios that you don't expect as a way to help catch bugs in other subsystems seems perfectly valid. And going to extra effort to support something that never happens in itself is usually a sign of working too hard. :) I merely sent out this PR as means to send simply, focused PRs that build on one another. I can hold off till we combine this with a later change, but the PR will be bigger -- though not huge, if the prototype I built is any indication.

**Comment by AArnott** [comment]
> I merely sent out this PR as means to send simply, focused PRs that build on one another.

Well, that, and to get early feedback such as yours to guide my next steps. :)

> If that now causes multiple changes in teh future that would break us (or anyone like us) that did not expect that to ever happen, and which we got communication would not happen.

Among the editor, Roslyn, and my team, we haven't decided exactly how the table manager will be exposed yet. I personally am currently in favor of Roslyn explicitly making it available. This approach should mitigate your concern I think. Any other change to the VS platform code to allow folks to access the table manager without the owner's approval would allow us to "light up" for more language services automatically, but also run straight into the problems you're predicting.

**Comment by CyrusNajmabadi** [comment]
Sounds good to me.

**Comment by CyrusNajmabadi** [comment]
> And as for waiting till then, sure

Note: i also mean "waiting" in an informal sense.  i.e. if you wanted to give me a quick sketch/rundown on how you think this would work, that woudl be fine as well.  I'm looking more for as large as understanding as possible (even if you can't share the strict feature you're working on).  With as much context as you can share, i can assess things as best as possible.  Thanks! :)

**Comment by AArnott** [comment]
>  I'm looking more for as large as understanding as possible (even if you can't share the strict feature you're working on). With as much context as you can share, i can assess things as best as possible.

Sounds good. I'm about to go OOF for a week anyway. So I'll probably get back to you after that, and after my manager gets back so I can clear exactly what I can disclose.

**Comment by AArnott** [comment]
Closing for now pending more work on my side.
